vendor/bat-native-ledger/src/bat/ledger/internal/request/post_connect/post_connect.cc

```diff
@@ -0,0 +1,136 @@
+             std::string::npos) {
+    // HTTP 403: Uphold
+    return type::Result::UPHOLD_TRANSACTION_VERIFICATION_FAILURE;
+  } else if (message->find("request signature verification failure") !=
+             std::string::npos) {
+    // HTTP 403: bitFlyer, Gemini
+    return type::Result::REQUEST_SIGNATURE_VERIFICATION_FAILURE;
+  } else if (message->find("unable to link - unusual activity") !=
+             std::string::npos) {
+    // HTTP 400: bitFlyer, Gemini, Uphold
+    return type::Result::FLAGGED_WALLET;
+  } else if (message->find("region not supported") != std::string::npos) {
+    // HTTP 400: bitFlyer, Gemini, Uphold
+    return type::Result::REGION_NOT_SUPPORTED;
+  } else if (message->find("mismatched provider account regions") !=
+             std::string::npos) {
+    // HTTP 400: bitFlyer, Gemini, Uphold
+    return type::Result::MISMATCHED_PROVIDER_ACCOUNT_REGIONS;
+  } else {
+    // bitFlyer, Gemini, Uphold
+    BLOG(0, "Unknown message!");
+    return type::Result::LEDGER_ERROR;
+  }
+}
+
+type::Result ProcessResponse(const type::UrlResponse& response) {
+  switch (response.status_code) {
+    case net::HTTP_BAD_REQUEST:
+      BLOG(0, "Invalid request");
+      return ParseBody(response.body);
+    case net::HTTP_FORBIDDEN:
+      BLOG(0, "Forbidden");
+      return ParseBody(response.body);
+    case net::HTTP_NOT_FOUND:
+      BLOG(0, "Not found");
+      return type::Result::NOT_FOUND;
+    case net::HTTP_CONFLICT:
+      BLOG(0, "Conflict");
+      return type::Result::DEVICE_LIMIT_REACHED;
+    case net::HTTP_INTERNAL_SERVER_ERROR:
+      BLOG(0, "Internal server error");
+      return type::Result::LEDGER_ERROR;
+    default:
+      if (response.status_code != net::HTTP_OK) {
+        BLOG(0, "Unexpected HTTP status: " << response.status_code);
+        return type::Result::LEDGER_ERROR;
+      } else {
+        return type::Result::LEDGER_OK;
+      }
+  }
+}
+
+}  // namespace
+
+// static
+void PostConnect::OnResponse(Callback callback,
+                             const type::UrlResponse& response) {
+  ledger::LogUrlResponse(__func__, response);
+  std::move(callback).Run(ProcessResponse(response));
+}
+
+PostConnect::PostConnect(LedgerImpl* ledger) : RequestBuilder(ledger) {}
+
+PostConnect::~PostConnect() = default;
+
+base::expected<std::string, base::Location> PostConnect::Url() const {
+  const auto wallet = ledger_->wallet()->GetWallet();
+  if (!wallet) {
+    BLOG(0, "Rewards wallet is null!");
+    return base::unexpected(base::Location::Current());
+  }
+
+  return ledger::endpoint::promotion::GetServerUrl(
+      base::StringPrintf(Path(), wallet->payment_id.c_str()));
+}
+
+base::expected<std::vector<std::string>, base::Location> PostConnect::Headers()
+    const {
+  const auto wallet = ledger_->wallet()->GetWallet();
+  if (!wallet) {
+    BLOG(0, "Rewards wallet is null!");
+    return base::unexpected(base::Location::Current());
+  }
+
+  const auto content = Content();
```

Comment:
Do we have to construct the body twice here? (Once when generating the body itself and once when generating the headers.) Is there a way that this can be made more efficient in the future?

Reply:
Sounds good!